src/Routing/PushState.purs

```diff
@@ -0,0 +1,241 @@
+  , makeInterface
+  , foldLocations
+  , locations
+  , foldPaths
+  , paths
+  , matches
+  , matchesWith
+  ) where
+
+import Prelude
+
+import Control.Monad.Eff (Eff)
+import Control.Monad.Eff.Ref (REF, modifyRef, newRef, readRef, writeRef)
+import DOM (DOM)
+import DOM.Event.EventTarget (addEventListener, eventListener) as DOM
+import DOM.HTML (window) as DOM
+import DOM.HTML.Event.EventTypes (popstate) as DOM
+import DOM.HTML.History (DocumentTitle(..), URL(..), pushState, replaceState, state) as DOM
+import DOM.HTML.Location (hash, pathname, search) as DOM
+import DOM.HTML.Types (HISTORY)
+import DOM.HTML.Types (htmlDocumentToDocument, windowToEventTarget) as DOM
+import DOM.HTML.Window (document, history, location) as DOM
+import DOM.Node.Document (createTextNode) as DOM
+import DOM.Node.MutationObserver (mutationObserver, observe) as DOM
+import DOM.Node.Node (setNodeValue) as DOM
+import DOM.Node.Types (textToNode) as DOM
+import Data.Array as Array
+import Data.Foldable (class Foldable, for_, indexl, traverse_)
+import Data.Foreign (Foreign)
+import Data.Maybe (Maybe(..), fromMaybe, maybe)
+import Data.Newtype (class Newtype, unwrap)
+import Data.StrMap as StrMap
+import Data.String as String
+import Routing (match)
+import Routing.Match (Match)
+
+type PushStateEffects eff =
+  ( history :: HISTORY
+  , dom :: DOM
+  , ref :: REF
+  | eff
+  )
+
+newtype Basename = Basename String
+
+derive newtype instance eqBasename :: Eq Basename
+derive newtype instance ordBasename :: Ord Basename
+derive instance newtypeBasename :: Newtype Basename _
+
+newtype Path = Path String
+
+derive newtype instance eqPath :: Eq Path
+derive newtype instance ordPath :: Ord Path
+derive instance newtypePath :: Newtype Path _
+
+-- | A `PushStateInterface` is a localized instance for making location changes
+-- | and consuming the events. Since the DOM API does not provide a general
+-- | event type for push state (only pop), you must use the coupled effects
+-- | along with `listen` to receive events on all location changes.
+-- |
+-- | * `pushState` – pushes a new location state and path onto the history stack.
+-- | * `replaceState` – replaces the location state and path in the history stack.
+-- | * `locationState` – Dereferences the current history state
+-- | * `listen` – Subscribes to location changes (both push and pop). Returns
+-- |   an effect which removes the listener.
+type PushStateInterface eff =
+  { pushState :: Foreign -> Path -> Eff eff Unit
+  , replaceState :: Foreign -> Path -> Eff eff Unit
+  , locationState :: Eff eff LocationState
+  , listen :: (LocationState -> Eff eff Unit) -> Eff eff (Eff eff Unit)
+  }
+
+-- | Creates a new `PushStateInterface` given a `Basename`. The `Basename` is
+-- | used as an implicit prefix for the paths supplied to the instance, being
+-- | automatically added and stripped as necessary.
+makeInterface :: forall eff. Basename -> Eff (PushStateEffects eff) (PushStateInterface (PushStateEffects eff))
```

Comment:
What's the motivation for having `Basename` be part of this lib? Personally I've always just found this kind of feature confuses more often than it helps. Maybe "enterprise apps" need it, but I'd rather just tell them to make a wrapper function around a simpler one.

Reply:
I think it's convenient. I'm not adamant it exists though. It's simple enough to write a transformation around `PushStateInstance` to layer it on after the fact as well.